### ISSUE TITLE
RTL support

### ISSUE THREAD

**Issue by jamma313** [Kozea/WeasyPrint#106]
Hi,

What is the current status of weasyprint concerning rtl support ?
It looks like it has a partial support (can display text but not boxes), is this a long way to finish it ?

Thanks.

**Comment by bahmany** [comment]
Hi, RTL support, Please

Thanks

**Comment by doctoromer** [comment]
Hey, any rtl support is planned soon?

**Comment by liZe** [comment]
> Hey, any rtl support is planned soon?

Adding RTL and BiDi support will be possible after rewriting the line breaking algorithm (#301), which is explained in [Fragmentation Module](https://www.w3.org/TR/css-break-3/#breaking-rules).

**That's huge.**

Secret breaking news: implementing the breaking rules and adding rtl/bidi support is on my short list for 2018. It'll come with dropping Python 2 support and releasing 1.0.0. Maybe.

**Comment by doctoromer** [comment]
Thank you very much! I will look forward to it.

**Comment by doctoromer** [comment]
thank you very much!

בתאריך יום ו׳, 23 במרץ 2018, 18:34, מאת mustafa0x ‏<notifications@github.com

**Comment by liZe** [comment]
> According to some brief tests with 0.42, RTL/BIDI supports seems to be pretty decent.

RTL suppot is not that bad for simple use cases and will probably be quite good in version 43.

Bidi is a much harder goal to reach. We'll have to open a separate issue when this one is closed.

**Comment by mohdnajdy** [comment]
@liZe Is there a way we can maybe chat or you can give me some guides on how to contribute to the RTL support ?

**Comment by dumblob** [comment]
@mohdnajdy did you have any success in adding support for RTL to weasyprint?

**Comment by mohdnajdy** [comment]
@dumblob not really. I am still lost on where to contribute at for supporting RTL

**Comment by liZe** [comment]
> I am still lost on where to contribute at for supporting RTL

It’s really hard to know where to start. `git grep rtl` will give a few TODOs, but that’s a small part of the work needed.

Some parts are probably easier to fix: tables, list markers, flex blocks… The hard work is of course in inline blocks layout.

You can find failing tests related to RTL in [WeasySuite](http://test.weasyprint.org/suite-css21/results/) (this page may take a long time to render). And, of course, we’ll need to write our own tests too.

Fixing RTL won’t mean that bidi is fixed. Supporting bidi would mean that we have our own implementation of the [bidi algorithm](http://www.unicode.org/reports/tr9/) and don’t rely on Pango’s one, as we have to handle line breaks with embedded tags. Yes, we would be able to drop Pango and use HarfBuzz instead, but that’s a lot of work…

**Comment by liZe** [comment]
> And, of course, we’ll need to write our own tests too.

:bulb: Writing failing tests is actually a really good starting point. :bulb:

**Comment by malnajdi** [comment]
Hello again @liZe I would like to re-state the issue of the Arabic text in a p tag. As shown in the picture. And i would really love to solve it. So please can you guide me to where i should look for to solve this issue ? if there is a simple solution please let me know.

![image](https://user-images.githubusercontent.com/15247545/94364389-84234180-00d1-11eb-8963-c0cd0c96c399.png)

**Comment by liZe** [comment]
> Hello again @liZe I would like to re-state the issue of the Arabic text in a p tag. As shown in the picture. And i would really love to solve it. So please can you guide me to where i should look for to solve this issue ? if there is a simple solution please let me know.

Hello again!

What’s wrong with this paragraph? I’m sorry, but as I don’t read Arabic it’s hard for me to spot the problem. If you have an HTML/CSS sample with the wrong rendering and an idea about what the right rendering is, it would be useful.

**Comment by grewn0uille** [comment]
I think the issue of the Arabic text in the p tag is that the third line starts too much on the right compared to the previous ones

**Comment by malnajdi** [comment]
> I think the issue of the Arabic text in the p tag is that the third line starts too much on the right compared to the previous ones

@liZe 
![issue_in_weasyprint](https://user-images.githubusercontent.com/15247545/94399564-46c1c100-0170-11eb-944f-eeb2bed3eeb6.png)

as you can see the first 2 lines are correct. The last line is going more to the right.

**Comment by liZe** [comment]
> as you can see the first 2 lines are correct. The last line is going more to the right.

Oh, OK 👓 !

I think that this problem is already fixed in master: see #1111.

**Comment by malnajdi** [comment]
@liZe great, Is there going to be a release date for the new version ?

**Comment by grewn0uille** [comment]
Hello!

If you’re interested in RTL and bidi support, we created a [short survey](https://app.evalandgo.com/s/index.php?id=JTk4cCU5NG8lOUQlQUQ%3D&a=JTk3biU5M2glOUMlQUI%3D) where you can give a boost to this feature and help us to improve WeasyPrint 😉

Vote for it!

**Comment by kesara** [comment]
As an argument for better RTL support, WeasyPrint is used by [IETF](https://www.ietf.org/) to generate PDFs of [RFCs](https://en.wikipedia.org/wiki/Request_for_Comments).
RTL support is essential in some RFCs.
For example, an internet draft on the way to becoming an RFC has issues with PDF rendering because of the WeasyPrint RTL support.
Ref: https://mailarchive.ietf.org/arch/msg/auth48archive/_8ZjoHcPh5HfYxRQb3P4zVbdXUg/

**Comment by grewn0uille** [comment]
Hello!

As it’s soon our 2-year anniversary as CourtBouillon, we opened a [short survey](https://thym.courtbouillon.org/1) to know more about your expectations.
Don’t hesitate to support this feature and give it a boost 🚀!

The survey will be opened until October 10th.

**Comment by grewn0uille** [comment]
Hello!

As you may know, two weeks ago was [CourtBouillon 3-year anniversary](https://www.courtbouillon.org/blog/00046-three-years-of-courtbouillon) 🎉.

For this occasion, we prepared [a short survey](https://thym.courtbouillon.org/4) to have your opinion on this year’s features and to know what you’d like to see in the future!
Don’t hesitate to give a boost to this feature ✨️

 The survey is opened until November 19.

**Comment by tttalshaul** [comment]
+1 for bidi support please..

**Comment by liZe** [comment]
> +1 for bidi support please.

We all want bidi support, help is appreciated!

> but when I'm copying the text from the PDF to a notepad, some of the numbers are copied to the end of the sentence even that they exist in the beginning of it.

I’m not sure that it’s a problem in WeasyPrint then. If the layout is right but the copy-paste is wrong, it’s maybe because of the PDF reader.

If you have two simple PDF files with bidi, with the same layout but different copy-paste results, feel free to share so that we can study them.

**Comment by tttalshaul** [comment]
> I’m not sure that it’s a problem in WeasyPrint then. If the layout is right but the copy-paste is wrong, it’s maybe because of the PDF reader.
You are right, I've upgraded the web browser and seems it handles the copy-paste better (not perfect yet). Adobe reader handles this worse..

**Comment by NHLOCAL** [comment]
What about that? Have support. Very partially partial

**Comment by moreaki** [comment]
Three incremental SVG text PRs are ready for review:

- #2754 shapes simple SVG text chunks as a whole, so combining marks and ligatures are kept together.
- #2755 builds on #2754 and applies SVG `direction` to text layout and `text-anchor` start/end behavior for RTL text.
- #2756 builds on #2755 and applies simple SVG `unicode-bidi` values to whole text chunks.

These are small SVG-focused steps toward better RTL/bidi behavior, not a full fix for this issue. In particular, full SVG 2 mapping of per-character `x`, `y`, `dx`, `dy`, and `rotate` lists onto shaped visual glyph clusters after bidi reordering remains separate work.
